Usermod: Add SkyStrip weather forecast usermod

ksedgwic opened the pull request:
This usermod displays the weather forecast on several parallel LED strips.
It currently includes Cloud, Wind, Temperature, 24 Hour Delta, and TestPattern views.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added SkyStrip weather visualization usermod with multiple dynamic views for LED strips
  * Integrated OpenWeatherMap API for real-time and forecast weather data
  * Multiple visualization modes: Cloud, Wind, Temperature, Delta, and Test Pattern
  * JSON-based configuration for segment selection and display parameters

* **Documentation**
  * Added SkyStrip Interpretation Guide explaining HSV-based color mappings for each view

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ksedgwic committed:
Add SkyStrip weather forecast usermod

ksedgwic committed:
Fix CodeRabbit issues:

- Avoid leaving segments permanently frozen; restore previous freeze state
- Fix negative modulo in local-time marker logic
- Handle invalid segments safely
- Clarify duplicate "N" table row
- Fix country normalizeLocation bug
- Don't include the API key in logging
- Tighten wording; avoid hardcoding OWM plan limits
- Fix rate-limit reset logic
- Harden rate-limit check and keep informative logging
- Add network timeout to avoid long blocking on bad links
- Check non-2xx status codes before parsing
- Right-size JSON buffer per target
- Fix potential buffer overrun in emitSeriesMDHM
- Use float instead of double for DataPoint.value
- Restore seg.freeze after rendering (and avoid freezing on zero-length)
- use PROGMEM to reduce RAM use
- Sanitize HSV inputs
- Clamp HSV values in conversion to RGB
- Clarify SAFETY_DELAY_MS type
- Clamp and wrap HSV inputs to prevent overflow/underflow artifacts
- Normalize wind direction to [0,360) before mapping to hue
- fixed compiler warning about shadowed variables
- delay initial checkhistory after fetch

ksedgwic commented on usermods/usermod_v2_skystrip/cloud_view.cpp:
```diff
@@ -0,0 +1,218 @@
+  int len = end - start + 1;
+  if (len == 0)
+    return;
+
```
fixed

ksedgwic commented on usermods/usermod_v2_skystrip/open_weather_map_source.h:
```diff
@@ -0,0 +1,43 @@
+                      bool& invalidate_history) override;
+  const char* configKey() const override { return "OpenWeatherMap"; }
+
+  void composeApiUrl(char* buf, size_t len) const;
+  bool geocodeOWM(std::string const& rawQuery, double& lat, double& lon, int* outMatches = nullptr);
```
fixed

ksedgwic commented on usermods/usermod_v2_skystrip/rest_json_client.cpp:
```diff
@@ -0,0 +1,61 @@
+    client = &plainClient;
+  }
+
+  // Begin request
+  if (!http_.begin(*client, url)) {
+    http_.end();
+    DEBUG_PRINTLN(F("SkyStrip: RestJsonClient::getJson: trouble initiating request"));
+    return nullptr;
+  }
+  DEBUG_PRINTF("SkyStrip: RestJsonClient::getJson: free heap before GET: %u\n", ESP.getFreeHeap());
```
tried the timeout, but it hangs

ksedgwic commented on usermods/usermod_v2_skystrip/test_pattern_view.cpp:
```diff
@@ -0,0 +1,179 @@
+    s = values[1] / 100.f;
+    v = values[2] / 100.f;
+    return true;
+  }
```
fixed

ksedgwic committed:
Add missing header include

ksedgwic committed:
Fix base API url

ksedgwic committed:
Code review improvements:

- Don't define USERMOD_ID_SKYSTRIP globally
- Use signed loop index to match len type
- Expose/read a socket timeout; ensure it’s applied on WiFiClient before begin()
- Guard against oversized payloads vs. JSON doc capacity
- Avoid wraparound trick in RestJsonClient ctor; initialize via reset for clarity
- Make RestJsonClient non-copyable/non-movable
- Preserve tail data in SkyModel mergeSeries
- Fix wind-direction interpolation (circular wrap-around bug)

ksedgwic committed:
Improve documentation and comments:

- Polish README list formatting and examples for consistency
- Improve FAQ
- Mark variable as unused
- Improve comments to clarify semantics
- Improve temperature mapping documentation in the README

ksedgwic committed:
Use segment virtualization APIs instead of raw start/stop

ksedgwic commented on usermods/usermod_v2_skystrip/rest_json_client.h:
```diff
@@ -0,0 +1,38 @@
+  HTTPClient http_;
+  unsigned long lastFetchMs_;
+  DynamicJsonDocument doc_;
+};
```
done

ksedgwic commented on usermods/usermod_v2_skystrip/skymodel.cpp:
```diff
@@ -0,0 +1,169 @@
+  auto itCut = std::lower_bound(current.begin(), current.end(), cutoff,
+                                [](const DataPoint& dp, time_t t){ return dp.tstamp < t; });
+  current.erase(current.begin(), itCut);
+}
```
fixed

ksedgwic commented on usermods/usermod_v2_skystrip/util.h:
```diff
@@ -0,0 +1,162 @@
+inline bool estimateDirAt(const SkyModel &m, time_t t, double step,
+                          double &out) {
+  return estimateAt(m.wind_dir_forecast, t, step, out);
+}
```
fixed

DedeHai commented:
Thanks for your contribution, this is a really elaborate UM!
I did not check the code but saw the picture on Discord. You should add some fotos of it in action so others know what its all about. From the readme and FAQ it was not really clear what it does exactly. 
You also may want to consider using palettes instead of fixed color values to make it a bit more versatile/colorful for others, just a suggestion.
Also add restrictions to the platforms you tested this on if there is anything hardware related (flash use, ram use, number of outputs), see json file in https://github.com/wled/WLED/pull/4427 for an example on how to do that.

ksedgwic committed:
skystrip: Fix formatting issues

ksedgwic committed:
skystrip: Fix urlEncode bounds check to avoid premature truncation

ksedgwic committed:
skystrip: Avoid undefined behavior: cast to unsigned char for ctype checks

ksedgwic committed:
skystrip: Guard against truncated input in parseLatLon

ksedgwic committed:
skystrip: Print 64-bit-safe timemachine dt

ksedgwic commented on usermods/usermod_v2_skystrip/FAQ.md:
```diff
@@ -0,0 +1,109 @@
+| NW        | 180     | Cyan   |
+| N         | 240     | Blue   | (wraps around)
+
+Note: Hues wrap at 360°, so “N” repeats at the boundary.
```
fixed

ksedgwic commented on usermods/usermod_v2_skystrip/open_weather_map_source.cpp:
```diff
@@ -0,0 +1,438 @@
+#include "skymodel.h"
+#include "util.h"
+
+static constexpr const char* DEFAULT_API_BASE = "http://api.openweathermap.org";
+static constexpr const char * DEFAULT_API_KEY = "";
+static constexpr const char * DEFAULT_LOCATION = "";
+static constexpr const double DEFAULT_LATITUDE = 37.80486;
+static constexpr const double DEFAULT_LONGITUDE = -122.2716;
+static constexpr unsigned DEFAULT_INTERVAL_SEC = 3600;	// 1 hour
+
```
we can't currently use https due to resource constraints (not enough heap for the TLS handshake + parsing giant JSON response).  WONTFIX

ksedgwic commented on usermods/usermod_v2_skystrip/open_weather_map_source.cpp:
```diff
@@ -0,0 +1,438 @@
+  out = deg + min / 60.0 + sec / 3600.0;
+  if (neg) out = -out;
+  return true;
+}
```
fixed

ksedgwic committed:
skystrip: Fix lat/long round-off update bug

ksedgwic committed:
skystrip: Improve documentation / comments

DedeHai commented on usermods/usermod_v2_skystrip/readme.md:
```diff
@@ -27,3 +27,16 @@ Enter the latitude and longitude for the desired forecast. You can:
+- SkyStrip was developed/tested using the
+  [Athom esp32-based LED strip controller](https://www.athom.tech/blank-1/wled-esp32-rf433-music-addressable-led-strip-controller).
+- I use a display consisting of 4 parallel 1-meter long
+  [WS2815 strips](https://www.superlightingled.com/dc12v-ws2815-upgraded-ws2812b-1m-144-leds-individually-addressable-digital-led-strip-lights-dual-signal-wires-waterproof-dream-color-programmable-5050-rgb-flexible-led-ribbon-light-p-2134:fd57dd8a8ac1ee0e78f5493a35b28792.html)
```
this link may go out of existance, add what it actually is

ksedgwic committed:
skystrip: Further improve documentation

ksedgwic committed:
skystrip: Even further doc improvements

DedeHai commented on usermods/usermod_v2_skystrip/readme.md:
```diff
@@ -0,0 +1,45 @@
+  strip of similar density should work; adjust power and wiring accordingly.
+- Based on comparisons with a baseline build SkyStrip uses:
+  - RAM: +2080 bytes
+  - Flash: +153,812 bytes
```
you are probably aware of this, but 150k is A LOT. Just know this may stop compiling at one point and you will get support questions :)

ksedgwic committed:
skystrip: Fix segment freeze semantics

ksedgwic committed:
skystrip: Don't gate fetching with strip.isUpdating

ksedgwic committed:
skystrip: Fix washed out areas with minimum saturation increase

ksedgwic committed:
skystrip: Suppress view of calm wind

ksedgwic committed:
skystrip: improve rendering of views

- formatting
- improve rain rendering in low PoP situations
- add precip rate (in/hr) to data collection
- render rain with PoP as saturation and in/hr as value
- tuned max delta value down

ksedgwic committed:
skystrip: display the SKYSTRIP_GIT_DESCRIBE in the config

ksedgwic committed:
skystrip: more view rendering improvements

- improve partial cloudiness depiction w/ dithering
- simplify and improve delta view
- refine temperature view, add new config color map
- use distinguished temp crossings

ksedgwic committed:
skystrip: fix location update reload bug

ksedgwic committed:
skystrip: conditional-compile remove the 10 second safety

This no longer needed, I think the base WLED does a safety revert

ksedgwic committed:
skystrip: improve the FAQ

ksedgwic committed:
skystrip: add explainer
